File: captainhcg/go/34-search-for-a-range.go
func searchRange(nums []int, target int) []int {
    return []int{findLeft(nums, 0, len(nums)-1, target), findRight(nums, 0, len(nums)-1, target)}
}


func findLeft(nums []int, st, ed, t int) int {
    if st > ed{
        return -1
    }
    mid := (st + ed)/2
    if nums[mid] < t{
        return findLeft(nums, mid+1, ed, t)
    } else if nums[mid] > t{
        return findLeft(nums, st, mid-1, t)
    } else {
        if (mid == 0) || (mid > 0 && nums[mid-1] != t){
            return mid
        }
        return findLeft(nums, st, mid-1, t)
    }
}

func findRight(nums []int, st, ed, t int) int {
    if st > ed{
        return -1
    }
    mid := (st + ed)/2
    if nums[mid] < t{
        return findRight(nums, mid+1, ed, t)
    } else if nums[mid] > t{
        return findRight(nums, st, mid-1, t)
    } else {
        if (mid == len(nums) - 1) || (mid < len(nums) - 1 && nums[mid+1] != t){
            return mid
        }
        return findRight(nums, mid+1, ed, t)
    }
}

File: bwu/binary_search/34-search-for-a-range.py
class Solution(object):
    def searchRange(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """
        n = len(nums)
        if not n:
            return [-1, -1]
            
        left, right = -1, -1
        
        l, r = 0, n - 1
        while l + 1 < r:
            mid = (l + r) / 2
            if nums[mid] < target:
                l = mid
            else:
                r = mid
        if nums[l] == target:
            left = l
        elif nums[r] == target:
            left = r
            
        l, r = 0, n - 1
        while l + 1 < r:
            mid = (l + r) / 2
            if nums[mid] > target:
                r = mid
            else:
                l = mid
        if nums[r] == target:
            right = r
        elif nums[l] == target:
            right = l
            
        return [left, right]
                
